src/ucs/datastruct/callbackq.c

```diff
@@ -247,3 +249,70 @@ ucs_status_t ucs_callbackq_remove_safe(ucs_callbackq_t *cbq, ucs_callback_t cb,
     ucs_callbackq_leave(cbq);
     return UCS_OK;
 }
+
+/* called with lock held */
+static ucs_callbackq_slow_elem_t* UCS_F_MAYBE_UNUSED
+ucs_callbackq_slow_path_find(ucs_callbackq_t *cbq,
+                             ucs_callbackq_slow_elem_t *elem)
+{
+    ucs_callbackq_slow_elem_t *it_elem;
+
+    ucs_queue_for_each(it_elem, &cbq->slow_path, queue){
+        if (it_elem == elem) {
+            return elem;
+        }
+    }
+    return NULL;
+}
+
+static void ucs_callbackq_slow_path_cb(void *arg)
+{
+    ucs_callbackq_t *cbq = arg;
+    ucs_callbackq_slow_elem_t *elem;
+    ucs_queue_iter_t iter;
+
+    ucs_callbackq_enter(cbq);
+
+    /* TODO avoid locks on every iteration. For instance implement
+     * lazy remove with additional list of elements to be removed. */
+    ucs_queue_for_each_safe(elem, iter, &cbq->slow_path, queue) {
```

Comment:
I guess it should, since we take the lock prior to update condition (which should be updated in the end of the cycle body)